### PR TITLE
Feat/directions

### PR DESCRIPTION
**Pull request by lyesehili**
Add Direction Display UI Feature

## 📋 Description

This PR implements the **UI layer for displaying directions** between the user's location and event pins on the map. The feature is designed with a clean, modular architecture that will make backend integration (Mapbox Directions API) straightforward next week.

### What's Implemented

- ✅ **Direction State Management** - MVVM architecture with DirectionViewModel
- ✅ **Visual Route Display** - Blue line with endpoint markers
- ✅ **UI Integration** - Direction icon button in event detail sheet
- ✅ **State Handling** - Clear state transitions (Cleared → Loading → Displayed)
- ✅ **Comprehensive Tests**

### User Flow

1. User taps on an event pin → Event detail sheet appears
2. User taps the **direction icon** (📍➡️) next to event title
3. Loading state shows briefly (500ms)
4. **Blue line appears** on map from user location to event
5. Icon turns red - user can tap again to clear directions
6. Directions auto-clear when closing event detail

## 📸 Screenshots

<p align="center">
  
  &nbsp;&nbsp;
  
</p>

### 🆕 New Files Created (7)

**Core Implementation:**
`DirectionState.kt` — Sealed class for direction states (Cleared/Loading/Displayed)  
`DirectionViewModel.kt` — ViewModel managing direction state (MVVM)  
`DirectionOverlay.kt` — Composable rendering the blue line on map  

**Tests:**
`DirectionStateTest.kt`  
`DirectionViewModelTest.kt`  
`MapScreenViewModelDirectionTest.kt`  
`EventDetailSheetDirectionTest.kt`  

### ✏️ Modified Files (4)

`EventDetailSheet.kt` — Added direction icon button next to event title  
`MapScreenViewModel.kt` — Integrated DirectionViewModel, added toggleDirections()  
`MapScreen.kt` — Integrated DirectionOverlay in map layers  
`UiTestTags.kt` — Added GET_DIRECTIONS_BUTTON test tag  

## 🧪 Testing

- ✅ **DirectionState**: 11 tests - all state transitions and edge cases
- ✅ **DirectionViewModel**: 16 tests - async operations, state management
- ✅ **Integration**: 8 tests - ViewModel integration with MapScreen
- ✅ **UI**: 13 tests - button visibility, clicks, content description

All tests pass successfully. Tests use proper coroutine testing with `StandardTestDispatcher`.

## 🔌 Backend Integration (Next Week)

The architecture is designed for easy backend integration:

### Known Limitation
The direction line color may be slightly affected by map theme (dark/light mode) due to Mapbox Standard Style's global transformations. This is a Mapbox limitation - all layers within MapboxStandardStyle are transformed when `lightPreset` changes. The line remains clearly visible and blue in both modes.

### Current Behavior
- **Origin**: Default EPFL location (will use GPS next week)
- **Route**: Straight line with 11 interpolated points (will use API route next week)
- **Performance**: Smooth rendering with no lag

## ✅ Checklist

- [x] Code follows project conventions
- [x] MVVM architecture respected
- [x] All tests pass
- [x] No compilation errors or warnings (except unused imports)
- [x] UI is responsive and smooth
- [x] Documentation is complete
- [x] Ready for backend integration

**Note**: This PR focuses entirely on UI. Backend integration with Mapbox Directions API is planned for next week's sprint.

## Summary by CodeRabbit

* **New Features**
  * Added "Get Directions" control to event details and a styled map overlay showing route and start/end markers; directions clear when closing details.
* **Tests**
  * New UI and unit tests covering the directions button visibility, labels, click behavior, rapid clicks, ordering, state toggling, plus overlay and view-model state lifecycle and rendering scenarios.

### CODE REVIEW

File: app/src/main/java/com/swent/mapin/ui/map/DirectionViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  private fun createStraightLineRoute(start: Point, end: Point): List<Point> {
+    val points = mutableListOf<Point>()
+    val steps = 10
+
+    for (i in 0..steps) {
+      val fraction = i.toDouble() / steps
+      val lat = start.latitude() + (end.latitude() - start.latitude()) * fraction
+      val lng = start.longitude() + (end.longitude() - start.longitude()) * fraction
+      points.add(Point.fromLngLat(lng, lat))
+    }
```
Comment (naderbahri) [inline app/src/main/java/com/swent/mapin/ui/map/DirectionViewModel.kt]: Nitpick: The interpolation algorithm assumes linear latitude/longitude monotonicity (latitude and longitude always increasing or decreasing). This can produce incorrect ordering for routes that go west-to-east or south-to-north, and permits no handling of wrapping at the International Date Line. Consider computing interpolated points by interpolating each coordinate independently while accounting for sign and wrap, or use a proper geodesic interpolation (great-circle) for correctness across larger distances and for routes that cross 180/-180 longitude. Suggestion: Replace simple linear interpolation with a geodesic interpolation helper or at least handle longitude wrap and non-monotonic direction. Add a small comment indicating this is intentionally simplified for demo and link future TODO to Mapbox Directions integration.

File: app/src/main/java/com/swent/mapin/ui/map/DirectionViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+   */
+  private fun createStraightLineRoute(start: Point, end: Point): List<Point> {
+    val points = mutableListOf<Point>()
+    val steps = 10
+
+    for (i in 0..steps) {
+      val fraction = i.toDouble() / steps
+      val lat = start.latitude() + (end.latitude() - start.latitude()) * fraction
+      val lng = start.longitude() + (end.longitude() - start.longitude()) * fraction
+      points.add(Point.fromLngLat(lng, lat))
```
Comment (naderbahri) [inline app/src/main/java/com/swent/mapin/ui/map/DirectionViewModel.kt]: Nitpick: The code constructs lat and lng, then calls Point.fromLngLat(lng, lat). This is correct but easy to misread. Add an inline comment clarifying the parameter order to avoid future accidental swaps. Consider using a factory helper like Point.from(lng = ..., lat = ...) or a small local function makePoint(lng, lat) to improve readability.

File: app/src/main/java/com/swent/mapin/ui/map/DirectionOverlay.kt [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+      rememberGeoJsonSourceState(key = "direction-line-source") { data = GeoJSONData(lineString) }
+
+  val endpointsSource =
+      rememberGeoJsonSourceState(key = "direction-endpoints-source") {
+        data =
+            GeoJSONData(
+                listOf(
+                    com.mapbox.geojson.Feature.fromGeometry(routePoints.first()),
+                    com.mapbox.geojson.Feature.fromGeometry(routePoints.last())))
+      }
```
Comment (naderbahri) [inline app/src/main/java/com/swent/mapin/ui/map/DirectionOverlay.kt]: - Nitpick: The GeoJSON endpoints source uses Feature.fromGeometry(routePoints.first()) and last(). If routePoints contains a single point, both features will be identical; add a short guard or comment explaining the intended behavior for single-point routes. If a single endpoint should show only one marker, deduplicate features when routePoints.size == 1.
- Suggestion: Use routePoints.distinctBy { it.toJson() } or a conditional to avoid duplicate features when the origin equals the destination.

File: app/src/main/java/com/swent/mapin/ui/map/DirectionOverlay.kt [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    lineWidth = DoubleValue(10.0)
+    lineOpacity = DoubleValue(0.3)
+    lineBlur = DoubleValue(5.0)
+  }
+
+  LineLayer(sourceState = directionSource, layerId = "direction-line-layer") {
+    lineColor = ColorValue(mainBlue)
+    lineWidth = DoubleValue(5.0)
+    lineOpacity = DoubleValue(0.95)
+  }
```
Comment (naderbahri) [inline app/src/main/java/com/swent/mapin/ui/map/DirectionOverlay.kt]: - Nitpick: Hard-coded color constants are fine for initial styling but reduce themeability and may conflict with dark map styles or user theme. Consider exposing color tokens or using theme / style resources so that map overlay colors can be adapted globally later, or at least document that map style may alter appearance. 
- Suggestion: Add TODO and/or allow passing colors as optional parameters for easier testing and theming.

File: app/src/main/java/com/swent/mapin/ui/map/DirectionOverlay.kt [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+ *
+ * The visual style is designed to match Google Maps' direction lines while maintaining consistency
+ * with modern map UI patterns.
+ *
+ * @param routePoints Ordered list of geographic points forming the route path. Must contain at
+ *   least one point. First point is the origin, last point is the destination.
+ * @note This uses Mapbox's MapboxDelicateApi for low-level layer management. Colors may be affected
+ *   by map theme (dark/light mode) due to Mapbox Standard Style's global transformations.
+ */
+@OptIn(MapboxDelicateApi::class)
```
Comment (naderbahri) [inline app/src/main/java/com/swent/mapin/ui/map/DirectionOverlay.kt]: Nitpick: Using MapboxDelicateApi requires caution. Add a comment explaining why the delicate API is required and what aspects might break with Mapbox upgrades. Include a short TODO for revisiting after migrating to a stable API surface.

File: app/src/main/java/com/swent/mapin/ui/map/MapScreen.kt [hunk 6]
```diff
@@ -408,6 +411,12 @@ private fun MapLayers(
     CreateHeatmapLayer(heatmapSource)
   }
 
+  // Render direction overlay if directions are displayed
+  val directionState = viewModel.directionViewModel.directionState
+  if (directionState is DirectionState.Displayed) {
+    DirectionOverlay(routePoints = directionState.routePoints)
+  }
```
Comment (naderbahri) [inline app/src/main/java/com/swent/mapin/ui/map/MapScreen.kt]: -  Nitpick: The DirectionOverlay is conditionally rendered from MapLayers using viewModel.directionViewModel.directionState. This couples UI map layer composition to a nested ViewModel instance stored on MapScreenViewModel. Consider passing only the routePoints into the MapLayers composable instead of exposing whole state, to reduce coupling and make MapLayers easier to test. 
- Suggestion: Change MapLayers signature to accept routePoints: List<Point>? and render if not null. Keep DirectionState handling at MapScreen level.

**Review by naderbahri — APPROVED** [review]
I approve the intent of the PR and the addition of the direction features.
Main improvements to address:

- Replace linear interpolation with geodesic interpolation or handle longitude wrapping (International Date Line).
- 
- Extract the interpolation point count constant to avoid test fragility.
- 
- Make GeoJSON colors and keys configurable or properly documented.
- 
- Externalize user-facing strings into resources for localization.
- 
- Inject DirectionViewModel via constructor or dependency injection to improve testability.
- 
- Reduce coupling by passing only routePoints to MapLayers.

Strengthen unit tests to cover edge cases (IDL, identical points) and minimize or mark flaky instrumented tests that depend on Mapbox.

Try to apply these changes to enhance the robustness and maintainability of the code.
